src/battery-check/battery-check.c

```diff
@@ -0,0 +1,141 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+
+#include <errno.h>
+#include <getopt.h>
+#include <sys/socket.h>
+#include <unistd.h>
+
+#include "battery-util.h"
+#include "build.h"
+#include "constants.h"
+#include "errno-util.h"
+#include "glyph-util.h"
+#include "fd-util.h"
+#include "log.h"
+#include "main-func.h"
+#include "socket-util.h"
+#include "terminal-util.h"
+
+static void help(void) {
+        printf("%s\n\n"
+               "Checks battery level to see whether there's enough charge.\n\n"
+               "   -h --help            Show this help\n"
+               "      --version         Show package version\n",
+               program_invocation_short_name);
+}
+
+static void battery_check_send_plymouth_message(char *message) {
+        assert(message);
+
+        ssize_t w;
+        static const union sockaddr_union sa = PLYMOUTH_SOCKET;
+        _cleanup_close_ int fd = -EBADF;
+        _cleanup_free_ char *plymouth_message = NULL, *plymouth_shutdown = NULL;
+
+        if (asprintf(&plymouth_message, "M\002%c%s", (int) strlen(message), message) < 0) {
+                log_oom();
+                return;
+        }
+        if (asprintf(&plymouth_shutdown, "C\002%c%s", (int) strlen("shutdown"), "shutdown") < 0) {
+                log_oom();
+                return;
+        }
+
+        /* We set SOCK_NONBLOCK here so that we rather drop the
+         * message than wait for plymouth */
+        fd = socket(AF_UNIX, SOCK_STREAM|SOCK_CLOEXEC|SOCK_NONBLOCK, 0);
+        if (fd < 0) {
+                log_warning_errno(errno, "socket() failed: %m");
+                return;
+        }
+
+        if (connect(fd, &sa.sa, SOCKADDR_UN_LEN(sa.un)) < 0) {
+                log_full_errno(IN_SET(errno, EAGAIN, ENOENT) || ERRNO_IS_DISCONNECT(errno) ? LOG_DEBUG : LOG_WARNING, errno, "Connection to plymouth failed: %m");
+                return;
+        }
+
+        w = strlen(plymouth_message);
+        if (write(fd, plymouth_message, strlen(plymouth_message)) != w)
+                log_full_errno(IN_SET(errno, EAGAIN, ENOENT) || ERRNO_IS_DISCONNECT(errno) ? LOG_DEBUG : LOG_WARNING, errno, "Failed to write to plymouth: %m");
```

Comment:
so this is actually not right.

in theory a socket can accept only a part of your write. which is something we are not expecting here (the data is just too small, the socket buffer too large), but we should check for this. this means write() can return negative (in which case errno is set) or can return positive but short of what we wanted write (in which case errno is not set).

hence you cannot rely on errno being set in the log_full_errno() parameters... hence what about:

```c
ssize_t n;
n = write(fd, plymouth_message, c);
if (n < 0) {
        log_full_errno(IN_SET(errno, EAGAIN, ENOENT) || ERRNO_IS_DISCONNECT(errno) ? LOG_DEBUG : LOG_WARNING, errno, "Failed to write to plymouth, ignoring: %m");
        return;
} 
if (n != c) {
        log_debug("Short write to plymouth, ignoring: %m");
        return;
}
```

Reply:
alternatively you can just use our helper:

```c
r = loop_write(fd, plymouth_message, c, /* do_pool= */ false);
if (r < 0)
        return (void) log_full_errno(IN_SET(r, -EAGAIN, -ENOENT) || ERRNO_IS_DISCONNECT(r) ? LOG_DEBUG : LOG_WARNING, r, "Failed to write to plymouth, ignoring: %m");
```

which is probably the safer bet. It loops around write() and if it cann't anymore returns EIO, i.e. turns short writes into proper errors.

Reply:
these two examples assume you store the return value of the asprintf() earlier in a variable `c`, as my earlier suggestion does. That way you can avoid the extra strlen()